scipy/signal/_signaltools.py

```diff
@@ -2893,12 +2898,13 @@ def envelope(z: np.ndarray, bp_in: tuple[int | None, int | None] = (1, None), *,
                          f"for n={z.shape[axis]=} and {bp_in=}!")
 
     # moving active axis to end allows to use `...` for indexing:
-    z = np.moveaxis(z, axis, -1)
+    z = xp.moveaxis(z, axis, -1)
 
-    if np.iscomplexobj(z):
+    if xp.isdtype(z.dtype, 'complex floating'):
         Z = sp_fft.fft(z)
     else:  # avoid calculating negative frequency bins for real signals:
-        Z = np.zeros_like(z, dtype=sp_fft.rfft(z.flat[:1]).dtype)
+        dt = sp_fft.rfft(xp.reshape(z, (-1,))[:1]).dtype
```

Comment:
Another style comment: What do you think about
```suggestion
        dt = sp_fft.rfft(xp[..., :1]).dtype
```
It should work due `z = xp.moveaxis(z, axis, -1)`.